File: browser/src/js/moneysocket/beacon/shared_seed.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

const Crypto = require('crypto');
const BinUtl = require('../utl/bin.js').BinUtl;

const SHARED_SEED_LEN = 16;

class SharedSeed {
    constructor(seed_bytes) {
        if (seed_bytes == null) {
            this.seed_bytes = Crypto.randomBytes(SHARED_SEED_LEN);
        } else {
            this.seed_bytes = seed_bytes;
        }
    }

    static fromHexStr(hex_str) {
        if (hex_str.length != SHARED_SEED_LEN * 2) {
            return null;
        }
        return new SharedSeed(BinUtl.toByteArray(hex_str));
    }

    toString() {
        return BinUtl.toHexString(this.seed_bytes);
    }

    getBytes() {
        return this.seed_bytes;
    }

    sha256(input_bytes) {
        const hash = Crypto.createHash('sha256');

        hash.update(input_bytes);
        return hash.digest();
    }

    doubleSha256(input_bytes) {
        return this.sha256(this.sha256(input_bytes));
    }

    deriveAes256Key() {
        return this.doubleSha256(this.seed_bytes);
    }

    deriveRendezvousId() {
        var aes256_key = this.deriveAes256Key();
        return this.doubleSha256(aes256_key);
    }

    deriveRendezvousIdHex() {
        return BinUtl.b2h(this.deriveRendezvousId());
    }
}

exports.SharedSeed = SharedSeed;

File: python/moneysocket/beacon/shared_seed.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

import os
import hashlib


class SharedSeed():
    SHARED_SEED_LEN = 16

    def __init__(self, seed_bytes=None):
        self.seed_bytes = (seed_bytes if seed_bytes is not None else
                           os.urandom(SharedSeed.SHARED_SEED_LEN))
        assert len(self.seed_bytes) == SharedSeed.SHARED_SEED_LEN

    @staticmethod
    def from_hex_string(hex_str):
        if len(hex_str) != SharedSeed.SHARED_SEED_LEN * 2:
            return None
        return SharedSeed(seed_bytes=bytes.fromhex(hex_str))

    def __str__(self):
        return self.seed_bytes.hex()

    def get_bytes(self):
        return self.seed_bytes

    def sha256(self, input_bytes):
        return hashlib.sha256(input_bytes).digest()

    def double_sha256(self, input_bytes):
        return self.sha256(self.sha256(input_bytes))

    def derive_aes256_key(self):
        return self.double_sha256(self.seed_bytes)

    def derive_rendezvous_id(self):
        aes256_key = self.derive_aes256_key()
        return self.double_sha256(aes256_key)

